pkg_auto/impl/util.sh

```diff
@@ -0,0 +1,253 @@
+    path=${1}; shift
+    dir_var_name=${1}; shift
+    local -n dir_ref=${dir_var_name}
+
+    if [[ -z ${path} ]]; then
+        dir_ref='.'
+        return 0
+    fi
+    local cleaned_up dn
+    # strip trailing slashes
+    cleaned_up=${path%%*(/)}
+    # strip duplicated slashes
+    cleaned_up=${cleaned_up//+(\/)/\/}
+    # strip last component
+    dn=${cleaned_up%/*}
+    if [[ -z ${dn} ]]; then
+        dir_ref='/'
+        return 0
+    fi
+    if [[ ${cleaned_up} = "${dn}" ]]; then
+        dir_ref='.'
+        return 0
+    fi
+    # shellcheck disable=SC2034 # it's a reference to external variable
+    dir_ref=${dn}
+}
+
+# Works like basename, but without spawning new processes.
+#
+# Params:
+#
+# 1 - path to operate on
+# 2 - name of a variable which will contain a basename of the path
+function basename_out() {
+    local path base_var_name
+    path=${1}; shift
+    base_var_name=${1}; shift
+    local -n base_ref=${base_var_name}
+
+    if [[ -z ${path} ]]; then
+        base_ref=''
+        return 0
+    fi
+    local cleaned_up dn
+    # strip trailing slashes
+    cleaned_up=${path%%*(/)}
+    if [[ -z ${cleaned_up} ]]; then
+        base_ref='/'
+        return 0
+    fi
+    # strip duplicated slashes
+    cleaned_up=${cleaned_up//+(\/)/\/}
+    # keep last component
+    dn=${cleaned_up##*/}
+    # shellcheck disable=SC2034 # it's a reference to external variable
+    base_ref=${dn}
+}
+
+if [[ ${BASH_SOURCE[-1]##*/} = 'util.sh' ]]; then
+    THIS=${BASH}
+    basename_out "${THIS}" THIS_NAME
+    THIS_DIR=.
+else
+    THIS=${BASH_SOURCE[-1]}
+    basename_out "${THIS}" THIS_NAME
+    dirname_out "${THIS}" THIS_DIR
+fi
+
+THIS=$(realpath "${THIS}")
+THIS_DIR=$(realpath "${THIS_DIR}")
+dirname_out "${BASH_SOURCE[0]}" PKG_AUTO_IMPL_DIR
+PKG_AUTO_IMPL_DIR=$(realpath "${PKG_AUTO_IMPL_DIR}")
+# shellcheck disable=SC2034 # may be used by scripts sourcing this file
+PKG_AUTO_DIR=$(realpath "${PKG_AUTO_IMPL_DIR}/..")
+
+# Prints an info line.
+#
+# Params:
+#
+# @ - strings to print
+function info() {
+    printf '%s: %s\n' "${THIS_NAME}" "${*}"
+}
+
+# Prints info lines.
+#
+# Params:
+#
+# @ - lines to print
+function info_lines() {
+    printf '%s\n' "${@/#/"${THIS_NAME}: "}"
+}
+
+# Prints an info to stderr and fails the execution.
+#
+# Params:
+#
+# @ - strings to print
+function fail() {
+    info "${@}" >&2
+    exit 1
+}
+
+# Prints infos to stderr and fails the execution.
+#
+# Params:
+#
+# @ - lines to print
+function fail_lines() {
+    info_lines "${@}" >&2
+    exit 1
+}
+
+# Yells a message.
+#
+# Params:
+#
+# @ - strings to yell
+function yell() {
+    echo
+    echo '!!!!!!!!!!!!!!!!!!'
+    echo "    ${*}"
+    echo '!!!!!!!!!!!!!!!!!!'
+    echo
+}
+
+# Prints help. Help is taken from the lines prefixed with double
+# hashes in the top sourcer of this file.
+function print_help() {
+    if [[ ${THIS} != "${BASH}" ]]; then
+        grep '^##' "${THIS}" | sed -e 's/##[[:space:]]*//'
+    fi
+}
```

Comment:
I guessed you must have done this before I saw the function. "Oh you didn't... oh you did." :rofl: It's fine. :grin: